### PR TITLE
Kafka feature 3105 v5

### PR DESCRIPTION
**Pull request by CosmosSun**
This changeset ensures the kafka output ability for eve.
- A new output method convenient for data analysis.
- Configure with --enable-rdkafka option when you need outpout eve with kafka.
- Update eve kafka output conf doc

Continuation of #4093.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
 https://redmine.openinfosecfoundation.org/issues/3105
Describe changes:
- Kafka output method convenient for data analysis
- Update eve kafka output conf doc
-

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by norg** [comment]
Doc looks fine to me, a bit more explanation for new users might be helpful

**Comment by jryberg** [comment]
This is really awesome!! Is there anything blocking this merge?

**Comment by CosmosSun** [comment]
> This is really awesome!! Is there anything blocking this merge?
https://redmine.openinfosecfoundation.org/issues/3105#note-3
maybe i should solved all these possible cases

File: src/output-json.c [hunk 1]
```diff
@@ -1018,7 +1027,29 @@ OutputInitResult OutputJsonInitCtx(ConfNode *conf)
...
 #endif
+#ifdef HAVE_LIBRDKAFKA
+        else if (json_ctx->json_out == LOGFILE_TYPE_KAFKA) {
+            ConfNode *kafka_node = ConfNodeLookupChild(conf, "kafka");
+            if (!json_ctx->file_ctx->sensor_name) {
+                char hostname[1024];
+                gethostname(hostname, 1023);
+                json_ctx->file_ctx->sensor_name = SCStrdup(hostname);
+            }
+            if (json_ctx->file_ctx->sensor_name  == NULL) {
```
Comment (jlucovsky) [inline src/output-json.c]: nit: 2 blanks between `sensor_name` and `==`

File: src/util-log-kafka.c [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+/**
+ * \brief Message delivery report callback.
+ * Called once for each message.
+ */
+static void msg_delivered (rd_kafka_t *rk,
+			   void *payload, size_t len,
+			   int error_code,
+			   void *opaque, void *msg_opaque) 
+{
+    rk = rk;
```
Comment (jlucovsky) [inline src/util-log-kafka.c]: `lhs = rhs` ... what's this for? L108-L112

File: src/util-log-kafka.c [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    log_ctx->kafka_setup.topic_name = ConfNodeLookupChildValue(kafka_node, "topic");
+    partitions =  ConfNodeLookupChildValue(kafka_node, "partitions");
+    log_ctx->kafka_setup.partitions = atoi(partitions);
+
+    /*create kafka ctx*/        
+    rd_kafka_conf_t *conf;
+    rd_kafka_topic_conf_t *topic_conf;        
+    char tmp[16];        
+    char errstr[512];
+    kafka_ctx = (SCLogKafkaContext*) SCCalloc(1, sizeof(SCLogKafkaContext));
```
Comment (jlucovsky) [inline src/util-log-kafka.c]: Make sure `kafka_ctx` is freed on error paths/early exits.

File: src/util-log-kafka.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    if (kafka_ctx == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate kafka context");
+        exit(EXIT_FAILURE);
+    }
+    
+    conf = rd_kafka_conf_new();
```
Comment (jlucovsky) [inline src/util-log-kafka.c]: Make sure you free `conf` with `rd_kafka_topic_conf_destroy` on error paths/early exits.

File: src/util-log-kafka.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
+/* Copyright (C) 2007-2019 Open Information Security Foundation
```
Comment (jlucovsky) [inline src/util-log-kafka.c]: Nit: change date to just `2019`

File: src/util-log-kafka.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * \file
+ * File-like output for logging:  kafka
+ */
+#include "suricata-common.h"
+#include "util-log-kafka.h"
+#include "util-logopenfile.h"
+
+#ifdef HAVE_LIBRDKAFKA
```
Comment (jlucovsky) [inline src/util-log-kafka.c]: IMO, this guard should be moved prior to the first `#include` statement.

File: src/util-logopenfile.c [hunk 7]
```diff
@@ -450,6 +450,24 @@ SCConfLogOpenGeneric(ConfNode *conf,
...
         }
         log_ctx->type = LOGFILE_TYPE_REDIS;
+#endif
+#ifdef HAVE_LIBRDKAFKA
+    } else if (strcmp(filetype, "kafka") == 0) {
+        ConfNode *kafka_node = ConfNodeLookupChild(conf, "kafka");
+        if (!log_ctx->sensor_name) {
+            char hostname[1024];
+            gethostname(hostname, 1023);
+            log_ctx->sensor_name = SCStrdup(hostname);
```
Comment (jlucovsky) [inline src/util-logopenfile.c]: Where will `sensor_name` be freed if `SCConfLogOpenKafka` fails?

**Comment by victorjulien** [comment]
Please submit a new PR with the requested updated, thanks!

**Comment by victorjulien** [comment]
Btw @dbcfd how do you feel about this? Asking because of your https://github.com/dbcfd/surikafka project.

@CosmosSun why do you think this should be a part of Suricata itself instead of using something like https://github.com/dbcfd/surikafka ?

**Comment by CosmosSun** [comment]
> 
> 
> Btw @dbcfd how do you feel about this? Asking because of your https://github.com/dbcfd/surikafka project.
> 
> @CosmosSun why do you think this should be a part of Suricata itself instead of using something like https://github.com/dbcfd/surikafka ?
I have not seen surikafka before, so i add a similar feature.
It may provide a new output type directly.

**Comment by zhouhanAC** [comment]
You should use rd_kafka_poll after each call to rd_kafka_produce, or call rd_kafka_poll regularly, otherwise the VmRSS of the process will continue to grow, causing memory leaks:

int LogFileWriteKafka(void *lf_ctx, const char *string, size_t string_len)
{    
    LogFileCtx *log_ctx = lf_ctx;    
    SCLogKafkaContext *kafka_ctx = log_ctx->kafka;
    int partition = kafka_ctx->partition % (log_ctx->kafka_setup.partitions);
    
    if (rd_kafka_produce(kafka_ctx->rkt, partition,
			RD_KAFKA_MSG_F_COPY,
			/* Payload and length */
			(void *)string, string_len,
			/* Optional key and its length */
			NULL, 0,
			/* Message opaque, provided in
			 * delivery report callback as
			 * msg_opaque. */
			NULL) == -1) 
	{
		SCLogError(SC_ERR_KAFKA,
				"%% Failed to produce to topic %s "
				"partition %i: %s\n",
				log_ctx->kafka_setup.topic_name, partition,
				rd_kafka_err2str(
					rd_kafka_errno2err(errno)));
		/* Poll to handle delivery reports */
		rd_kafka_poll(kafka_ctx->rk, 0);
        return -1;
	}
    kafka_ctx->partition++;

    rd_kafka_poll(kafka_ctx->rk, 0 /*non-blocking*/);
    return 0;
}

**Comment by zhouhanAC** [comment]
You should use rd_kafka_poll after each call to rd_kafka_produce, or call rd_kafka_poll regularly, otherwise the VmRSS of the process will continue to grow, causing memory leaks:

`int LogFileWriteKafka(void *lf_ctx, const char *string, size_t string_len)
{    
    LogFileCtx *log_ctx = lf_ctx;    
    SCLogKafkaContext *kafka_ctx = log_ctx->kafka;
    int partition = kafka_ctx->partition % (log_ctx->kafka_setup.partitions);
    
    if (rd_kafka_produce(kafka_ctx->rkt, partition,
			RD_KAFKA_MSG_F_COPY,
			/* Payload and length */
			(void *)string, string_len,
			/* Optional key and its length */
			NULL, 0,
			/* Message opaque, provided in
			 * delivery report callback as
			 * msg_opaque. */
			NULL) == -1) 
	{
		SCLogError(SC_ERR_KAFKA,
				"%% Failed to produce to topic %s "
				"partition %i: %s\n",
				log_ctx->kafka_setup.topic_name, partition,
				rd_kafka_err2str(
					rd_kafka_errno2err(errno)));
		/* Poll to handle delivery reports */
		rd_kafka_poll(kafka_ctx->rk, 0);
        return -1;
	}
    kafka_ctx->partition++;

    rd_kafka_poll(kafka_ctx->rk, 0 /*non-blocking*/);
    return 0;
}`

**Comment by victorjulien** [comment]
@zhouhanAC please report bugs in our issue tracker instead of very old and closed PRs, thanks!

**Pull request closed without merging**